File: medium/Generate-Parentheses-22.cpp
// Need to rework and test

class Solution {
public:
    vector<string> vs;
    
    vector<string> generateParenthesis(int n) {
        if (n <=0)
            return vs;
        char s[n*2];
        
        generateParenthesis_int(n, n, s, 0);
        return vs;
        
    }
    void generateParenthesis_int(int left, int right, char *s, int index) {
        if (left < 0 || left > right)
            return;
        if (left ==0 && right == 0) {
            
            vs.push_back(s);
            
        } else {
        if (left > 0) {
            s[index] = '(';
            generateParenthesis_int(left-1, right, s, index+1);
        }
        if (right > left)
             generateParenthesis_int(left, right-1, s, index+1);
        }
        return;
    }
        
};

File: medium/Generate-Parentheses-22.java
class Solution {
    private void addParen(ArrayList<String> list, int leftRem,
        int rightRem, char[] str, int count) {
        if (leftRem < 0 || rightRem < leftRem) return; // invalid state

        if (leftRem == 0 && rightRem == 0) { /* no more parens left */
            String s = String.copyValueOf(str);
            list.add(s);
        } else {
            /* Add left paren, if there are any left parens remaining. */
            if (leftRem > 0) {
                str[count] = '(';
                addParen(list, leftRem - 1, rightRem, str, count + 1);
            }

            /* Add right paren., if expression is valid */
            if (rightRem > leftRem) {
                str[count] = ')';
                addParen(list, leftRem, rightRem - 1, str, count + 1);
            }
        }
    }

    public ArrayList<String> generateParenthesis(int count) {
        char[] str = new char[count*2];
        ArrayList<String> list = new ArrayList<String>();
        addParen(list, count, count, str, 0);
        return list;
    }
};
